app/controllers/AuthenticationController.scala

```diff
@@ -429,6 +431,91 @@ class AuthenticationController @Inject()(
+          value: Cookie <- combinedAuthenticatorService.init(authenticator)
+          result: AuthenticatorResult <- combinedAuthenticatorService.embed(value, Redirect("/dashboard"))
+          _ <- multiUserDAO.updateLastLoggedInIdentity(user._multiUser, user._id)(GlobalAccessContext)
+          _ = userDAO.updateLastActivity(user._id)(GlobalAccessContext)
+        } yield result
+      case None =>
+        Future.successful(BadRequest(Messages("error.noUser")))
+      case Some(_) => Future.successful(BadRequest(Messages("user.deactivated")))
+    }
+
+  // Is called after user was successfully authenticated
+  def loginOrSignupViaOidc(oidc: OpenConnectId): Request[AnyContent] => Future[Result] = {
+    implicit request: Request[AnyContent] =>
+      userService.userFromMultiUserEmail(oidc.email)(GlobalAccessContext).futureBox.flatMap {
+        case Full(user) =>
+          val loginInfo = LoginInfo("credentials", user._id.toString)
+          loginUser(user, loginInfo)
+        case Empty =>
+          for {
+            organization: Organization <- organizationService.findOneByInviteByNameOrDefault(None, None)(
+              GlobalAccessContext)
+            user: User <- userService.insert(organization._id,
+                                             oidc.email,
+                                             oidc.given_name,
+                                             oidc.family_name,
+                                             isActive = true,
+                                             userService.getOIDCPasswordInfo)
+            multiUser: MultiUser <- multiUserDAO.findOne(user._multiUser)(GlobalAccessContext)
+            _ = analyticsService.track(SignupEvent(user, hadInvite = false))
+            //  brainDBResult <- brainTracing.registerIfNeeded(user, signUpData.password).toFox // Is this necessary?
+            _ = if (conf.Features.isDemoInstance) {
+              mailchimpClient.registerUser(user, multiUser, tag = MailchimpTag.RegisteredAsUser)
+            } else {
+              Mailer ! Send(defaultMails.newUserMail(user.name, oidc.email, None, enableAutoVerify = true))
+            }
+            _ = Mailer ! Send(
+              defaultMails.registerAdminNotifyerMail(user.name, oidc.email, None, organization, autoActivate = true))
+            // After registering, also login
+            loginInfo = LoginInfo("credentials", user._id.toString)
+            loginResult <- loginUser(user, loginInfo)
+          } yield loginResult
+        case _ => Future.successful(InternalServerError)
+      }
+  }
+
+  def openIdCallback(): Action[AnyContent] = Action.async { implicit request =>
+    for {
+      code <- openIdClient.getToken(oidcConfig,
+                                    absoluteOidcCallbackURL,
+                                    request.queryString.get("code").flatMap(_.headOption).getOrElse("missing code"),
+      )
+      oidc: OpenConnectId <- validateJsValue[OpenConnectId](code).toFox
```

Comment:
```suggestion
      openConnectId <- validateJsValue[OpenConnectId](code).toFox
```
(if I understand correctly that OpenIdConnect uses OpenConnectIds? Seems confusing to me, but may be correct if that’s what the protocol states)

Reply:
This is not something that is directly part of the protocol. I renamed it to OpenIdConnectClaimSet since it describes it more accurately